src/material/dialog/dialog-container.ts

```diff
@@ -228,9 +228,17 @@ export abstract class _MatDialogContainerBase extends BasePortalOutlet {
   /** Returns whether focus is inside the dialog. */
   private _containsFocus() {
     const element = this._elementRef.nativeElement;
-    const activeElement = this._document.activeElement;
+    const activeElement = this._getActiveElement();
     return element === activeElement || element.contains(activeElement);
   }
+
+  /** Gets the currently-focused element on the page. */
+  private _getActiveElement(): Element | null {
+    // If the `activeElement` is inside a shadow root, `document.activeElement` will
+    // point to the shadow root so we have to descend into it ourselves.
+    const activeElement = this._document.activeElement;
+    return activeElement?.shadowRoot?.activeElement as HTMLElement || activeElement;
```

Comment:
Isn't it possible for the element to be within multiple layers of shadow dom?

Reply:
It is possible, but my assumption was `activeElement` would point to the closest shadow root. That being said, I haven't actually tried it. If that's the case, we might need some reusable utility under `cdk/platform` since we have a lot of places that look only at `document.activeElement`.